os/fs/smartfs/smartfs_utils.c

```diff
@@ -1015,251 +1076,261 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
-
-	if (sectorno == 0xFFFF) {
-		/* Allocate a new sector for the file / dir */
-
-		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-		if (ret < 0) {
-			goto errout;
+			RESET_ENTRY_FLAGS(&new_entry);
 		}
+		SET_ENTRY_FLAGS(new_entry.flags, mode, type);
+	}
 
-		nextsector = (uint16_t)ret;
-
-		/* Set the newly allocated sector's type (file or dir) */
-
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-		if (sf) {
-			/* Using sector buffer and we have an open file context.  Just update
-			 * the sector buffer in the open file context.
-			 */
-
-			memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
-			chainheader = (struct smartfs_chain_header_s *)sf->buffer;
-			chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
-		} else
-#endif
-		{
-			if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
-				chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
-			} else {
-				chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			}
-
-			readwrite.count = 1;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, type);
-			readwrite.buffer = (uint8_t *)&chainheader->type;
-			readwrite.logsector = nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error %d setting new sector type for sector %d\n", ret, nextsector);
-				goto errout;
-			}
-		}
+	if (newchain) {
+		memset(fs->fs_rwbuffer, CONFIG_SMARTFS_ERASEDSTATE, entrysize + sizeof(struct smartfs_chain_header_s));
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+		chainheader = (struct smartfs_chain_header_s *)&fs->fs_rwbuffer[0];
+		chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
 	} else {
-		/* Use the provided sector number */
-
-		nextsector = sectorno;
+		memset(fs->fs_rwbuffer, CONFIG_SMARTFS_ERASEDSTATE, entrysize);
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[0];
 	}
 
 	/* Create the directory entry to be written in the parent's sector */
 
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	smartfs_wrle16(&entry->firstsector, nextsector);
+	smartfs_wrle16(&entry->firstsector, new_entry.firstsector);
 	smartfs_wrle16(&entry->utc, time(NULL));
 #else
-	entry->firstsector = nextsector;
+	entry->firstsector = new_entry.firstsector;
 	entry->utc = time(NULL);
 #endif
 	memset(entry->name, 0, fs->fs_llformat.namesize);
 	strncpy(entry->name, filename, fs->fs_llformat.namesize);
+	entry->flags = new_entry.flags;
 
 	/* Now write the new entry to the parent directory sector */
-
-	readwrite.logsector = psector;
-	/* If it chained right before, we should update chainheader */
-	if (update_chain) {
-		memset(fs->fs_rwbuffer, CONFIG_SMARTFS_ERASEDSTATE, sizeof(struct smartfs_chain_header_s));
-		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
-		chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
-		readwrite.count = entrysize + sizeof(struct smartfs_chain_header_s);
+	if (newchain) {
```

Comment:
yes it also we don't need it. My old PR just changed sequence of chaing...not affect to this.

Reply:
With refactoring, your PR changes will also become outdated.